rust/chains/hyperlane-ethereum/src/mailbox.rs

```diff
@@ -307,111 +301,27 @@ where
 {
     #[instrument(skip(self))]
     async fn count(&self, maybe_lag: Option<NonZeroU64>) -> ChainResult<u32> {
-        let base_call = self.contract.count();
-        let call_with_lag = if let Some(lag) = maybe_lag {
-            let tip = self
-                .provider
-                .get_block_number()
-                .await
-                .map_err(ChainCommunicationError::from_other)?
-                .as_u64();
-            base_call.block(tip.saturating_sub(lag.get()))
-        } else {
-            base_call
-        };
-        let count = call_with_lag.call().await?;
-        Ok(count)
-    }
-
-    #[instrument(skip(self))]
-    async fn delivered(&self, id: H256) -> ChainResult<bool> {
-        Ok(self.contract.delivered(id.into()).call().await?)
-    }
-
-    #[instrument(skip(self))]
-    async fn latest_checkpoint(&self, maybe_lag: Option<NonZeroU64>) -> ChainResult<Checkpoint> {
-        let base_call = self.contract.latest_checkpoint();
-        let call_with_lag = match maybe_lag {
-            Some(lag) => {
-                let tip = self
-                    .provider
-                    .get_block_number()
-                    .await
-                    .map_err(ChainCommunicationError::from_other)?
-                    .as_u64();
-                base_call.block(tip.saturating_sub(lag.get()))
-            }
-            None => base_call,
-        };
-        let (root, index) = call_with_lag.call().await?;
-        Ok(Checkpoint {
-            mailbox_address: self.address(),
-            mailbox_domain: self.domain.id(),
-            root: root.into(),
-            index,
-        })
-    }
-
-    #[instrument(skip(self))]
-    #[allow(clippy::needless_range_loop)]
-    async fn tree(&self, lag: Option<NonZeroU64>) -> ChainResult<IncrementalMerkle> {
-        let lag = lag.map(|v| v.get()).unwrap_or(0).into();
-
-        // use consistent block for all storage slot or view calls to prevent
-        // race conditions where tree contents change between calls
-        let fixed_block_number = self
+        let lag = maybe_lag.map(|v| v.get()).unwrap_or(0).into();
+        let fixed_block_number: BlockNumber = self
```

Comment:
Can we actually move back to the `let base_call = self.contract.count();` pattern? the nice part of it is that if there isn't a lag configured, then we save an RPC, and the eth_call is just made against the latest block. This has proved nice in practice

The reason why we needed to have the `fixed_block_number` logic before always specify a block number is because we were getting the tree by making multiple eth_getStorageAt RPCs at a specific height. Now that we can just call `tree()` with a single RPC, we don't need to worry about making multiple calls accidentally at different heights, so we can go back to the pattent we use elsewhere

Reply:
good callout